Form Errors Reactivity & Nested Fields Errors

### Describe the bug

There is def. something weird with the errors and im not sure yet why, but ill leave this issue to follow up

- The first i noticed while investigating some time ago, is that errors behave in a weird way in hybridly ( and a bit different from inertia ), when the backend sends the same error twice for a specific field, the errors variable is not triggered again ( and makes sense since the message is the same, and so the value never changes, so it will not trigger the watch effect. I believe this is related to original issue, reason why i wanted to mention.

- Seems like when using nested fields in forms, if we submit and the field has an error, the error is correctly set and all good here, but if you attempt to fill the said field ( so it does not contain errors ), the form.errors for that specific field seems like its never removed, while the server sends back a response without the given error shared. Its expected that also nested fields are also cleared when they are not present.

- Please do note that is also breaks the watch effect, so if you have a v-if for the given error, it will always be shown, because the variable is never removed.
 
 The only way to proper workaround this is to use the before hook to clear the given fields, but i believe this should be looked properly to have a smooth expirience.
 
 
 To note also that nested fields errors must be accessed with `form.errors['foo.nested']` instead of dot notation, ( not 100% sure but i believe it would be a nice DX to provide nested navigation for those as well.
 

### Reproduction

https://github.com/nikuscs/hybridly-demo-validation-errors

### Steps to reproduce

- Visit the https://blue-bird.test/login
- First submit without the nested field so it the error is flashed
- Then fill the contents of nested field and send again
- Errors for the given field will not be sent back by the server, but still persisted in errors
- Watch console for proper errors.

### System information

```bash
System:
    OS: macOS 13.3.1
    CPU: (8) arm64 Apple M2
    Memory: 201.41 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.8.0 - ~/Library/Caches/fnm_multishells/68334_1681237407108/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.2.0 - ~/Library/Caches/fnm_multishells/68334_1681237407108/bin/npm
  Browsers:
    Brave Browser: 112.1.50.114
    Chrome: 111.0.5563.146
    Firefox: 110.0
    Safari: 16.4
```


### Used package manager

pnpm

### Logs

_No response_

### Validations

- [X] Read the [docs](https://hybridly.dev).
- [X] Check that there isn't [already an issue](https://github.com/hybridly/hybridly/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Hybridly issue and not an issue related to something else (Vite, Vue...). For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A, join our [Discord server](https://discord.gg/uZ8eC7kRFV).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

I could use some help linking my local copy of hybridly into the Laravel skeleton.

Using the following in my package.json does not seem to be enough

```json
"hybridly": "file:../hybridly/packages/hybridly/dist",
```

Starting Vite in dev mode results in the following error.

```
> @ dev /Users/nicolas/Contributing/issue-71
> vite

failed to load config from /Users/nicolas/Contributing/issue-71/vite.config.ts
error when starting dev server:
Error: Cannot find module 'hybridly/vite'
Require stack:
- /Users/nicolas/Contributing/issue-71/vite.config.ts
- /Users/nicolas/Contributing/issue-71/node_modules/.pnpm/vite@4.3.3_@types+node@18.16.1/node_modules/vite/dist/node/chunks/dep-a178814b.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1039:15)
    at Module._load (node:internal/modules/cjs/loader:885:27)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/nicolas/Contributing/issue-71/vite.config.ts:37:28)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at _require.extensions.<computed> [as .js] (file:///Users/nicolas/Contributing/issue-71/node_modules/.pnpm/vite@4.3.3_@types+node@18.16.1/node_modules/vite/dist/node/chunks/dep-a178814b.js:64432:24)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
 ELIFECYCLE  Command failed with exit code 1.
```

What is your setup when developing Hybridly @innocenzi ?

--- a/packages/core/src/router/types.ts
+++ b/packages/core/src/router/types.ts
@@ -238,6 +238,4 @@ export interface Progress {
 	percentage: Readonly<number>
 }
 
-export interface Errors {
-	[key: string]: string
-}
+export type Errors = any

--- a/packages/utils/src/utils.spec.ts
+++ b/packages/utils/src/utils.spec.ts
@@ -1,0 +1,37 @@
+import { describe, it, expect } from 'vitest'
+import { setValueAtPath, unsetPropertyAtPath } from './utils'
+
+describe('utils', () => {
+	it('sets a value at a path in an object', () => {
+		const obj = {}
+		setValueAtPath(obj, 'foo.bar.baz', 'qux')
+		expect(obj).toEqual({ foo: { bar: { baz: 'qux' } } })
+	})
+
+	it('unsets a property at a path in an object', () => {
+		const obj = {
+			foo: {
+				bar: 'bar',
+				baz: 'baz',
+			},
+		}
+		unsetPropertyAtPath(obj, 'foo.bar')
+		expect(obj).toEqual({
+			foo: {
+				baz: 'baz',
+			},
+		})
+	})
+
+	it('deletes empty objects along the way when unsetting a property at a path in an object', () => {
+		const obj = {
+			foo: {
+				bar: {
+					baz: 'baz',
+				},
+			},
+		}
+		unsetPropertyAtPath(obj, 'foo.bar.baz')
+		expect(obj).toEqual({})
+	})
+})

--- a/packages/utils/src/utils.ts
+++ b/packages/utils/src/utils.ts
@@ -86,3 +86,64 @@ export function merge<T>(x: Partial<T>, y: Partial<T>, options: MergeOptions = {
 export function removeTrailingSlash(string: string): string {
 	return string.replace(/\/+$/, '')
 }
+
+/**
+ * Sets a value at a path in an object
+ *
+ * This function will set a value at a path in an object, creating any missing
+ * objects along the way. The object is modified in place.
+ *
+ * @param obj the object to set the value in
+ * @param path a dot-separated path to the property to set
+ * @param value the value to set
+ */
+export function setValueAtPath(obj: any, path: string, value: any): void {
+	// If the path doesn't contain a dot, then we can just set the value.
+	if (!path.includes('.')) {
+		obj[path] = value
+		return
+	}
+
+	// Otherwise, we need to split the path into segments and walk down the
+	// object tree until we find the right place to set the value.
+	const segments = path.split('.')
+	let nestedObject = obj
+	for (let i = 0; i < segments.length - 1; i++) {
+		const key = segments[i]
+		nestedObject = nestedObject[key] = nestedObject[key] || {}
+	}
+	nestedObject[segments[segments.length - 1]] = value
+}
+
+/**
+ * Unsets a property at a path in an object
+ *
+ * This function will unset a property at a path in an object, deleting any
+ * objects along the way that are empty. The object is modified in place.
+ *
+ * @param obj the object to unset the property in
+ * @param path a dot-separated path to the property to unset
+ */
+export function unsetPropertyAtPath(obj: any, path: string): void {
+	// If the path doesn't contain a dot, then we can just delete the property.
+	if (!path.includes('.')) {
+		delete obj[path]
+		return
+	}
+
+	// Otherwise, we need to split the path into segments and walk down the
+	// object tree until we find the right place to delete the property.
+	const segments = path.split('.')
+	let nestedObject = obj
+	for (let i = 0; i < segments.length - 1; i++) {
+		const key = segments[i]
+		nestedObject = nestedObject[key] = nestedObject[key] || {}
+	}
+
+	delete nestedObject[segments[segments.length - 1]]
+
+	// If the nested object is now empty, delete it.
+	if (Object.keys(nestedObject).length === 0) {
+		unsetPropertyAtPath(obj, segments.slice(0, -1).join('.'))
+	}
+}

--- a/packages/vue/src/composables/form.ts
+++ b/packages/vue/src/composables/form.ts
@@ -2,7 +2,7 @@ import isEqual from 'lodash.isequal'
 import type { Progress, UrlResolvable, HybridRequestOptions } from '@hybridly/core'
 import type { DeepReadonly } from 'vue'
 import { computed, reactive, ref, toRaw, watch } from 'vue'
-import { clone } from '@hybridly/utils'
+import { clone, setValueAtPath, unsetPropertyAtPath } from '@hybridly/utils'
 import { router } from '@hybridly/core'
 import { state } from '../stores/state'
 
@@ -177,7 +177,7 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Clears all errors.
 	 */
-	function clearErrors(...keys: (keyof T)[]) {
+	function clearErrors(...keys: string[]) {
 		if (keys.length === 0) {
 			keys = Object.keys(fields)
 		}
@@ -201,18 +201,8 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Clears the given field's error.
 	 */
-	function clearError(key: keyof T) {
-		delete errors.value[key]
-	}
-
-	function setNestedValue(obj: any, path: string, value: any) {
-		const segments = path.split('.')
-		let nestedObject = obj
-		for (let i = 0; i < segments.length - 1; i++) {
-			const key = segments[i]
-			nestedObject = nestedObject[key] = nestedObject[key] || {}
-		}
-		nestedObject[segments[segments.length - 1]] = value
+	function clearError(key: string) {
+		unsetPropertyAtPath(errors.value, key)
 	}
 
 	/**
@@ -220,14 +210,8 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	 */
 	function setErrors(incoming: Record<string, string>) {
 		clearErrors()
-		Object.entries(incoming).forEach(([key, value]) => {
-			// if the key is a dot-notated path, we expand it to set the error
-			// on the correct field
-			if (key.includes('.')) {
-				setNestedValue(errors.value, key, value)
-			} else {
-				errors.value[key] = value
-			}
+		Object.entries(incoming).forEach(([path, value]) => {
+			setValueAtPath(errors.value, path, value)
 		})
 	}
 

--- a/packages/vue/src/composables/form.ts
+++ b/packages/vue/src/composables/form.ts
@@ -205,13 +205,29 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 		delete errors.value[key]
 	}
 
+	function setNestedValue(obj: any, path: string, value: any) {
+		const segments = path.split('.')
+		let nestedObject = obj
+		for (let i = 0; i < segments.length - 1; i++) {
+			const key = segments[i]
+			nestedObject = nestedObject[key] = nestedObject[key] || {}
+		}
+		nestedObject[segments[segments.length - 1]] = value
+	}
+
 	/**
 	 * Sets current errors.
 	 */
 	function setErrors(incoming: Record<string, string>) {
 		clearErrors()
 		Object.entries(incoming).forEach(([key, value]) => {
-			errors.value[key as keyof T] = value as any
+			// if the key is a dot-notated path, we expand it to set the error
+			// on the correct field
+			if (key.includes('.')) {
+				setNestedValue(errors.value, key, value)
+			} else {
+				errors.value[key] = value
+			}
 		})
 	}
 

--- a/packages/vue/src/composables/form.ts
+++ b/packages/vue/src/composables/form.ts
@@ -4,11 +4,17 @@ import type { DeepReadonly } from 'vue'
 import { computed, reactive, ref, toRaw, watch } from 'vue'
 import { clone, setValueAtPath, unsetPropertyAtPath } from '@hybridly/utils'
 import { router } from '@hybridly/core'
+import type { Path, SearchableObject } from '@clickbar/dot-diver'
+import { getByPath } from '@clickbar/dot-diver'
 import { state } from '../stores/state'
 
-type Fields = Record<string, any>
+type Errors<T extends SearchableObject> = {
+	[K in keyof T]?: T[K] extends Record<string, any>
+		? Errors<T[K]>
+		: string;
+}
 
-interface FormOptions<T extends Fields> extends Omit<HybridRequestOptions, 'data' | 'url'> {
+interface FormOptions<T extends SearchableObject> extends Omit<HybridRequestOptions, 'data' | 'url'> {
 	fields: T
 	url?: UrlResolvable | (() => UrlResolvable)
 	key?: string | false
@@ -29,14 +35,17 @@ interface FormOptions<T extends Fields> extends Omit<HybridRequestOptions, 'data
 	/**
 	 * Callback executed before the form submission for transforming the fields.
 	 */
-	transform?: (fields: T) => Fields
+	transform?: (fields: T) => any
 }
 
 function safeClone<T>(obj: T): T {
 	return clone(toRaw(obj))
 }
 
-export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
+export function useForm<
+	T extends SearchableObject,
+	P extends Path<T> & string = Path<T> & string
+>(options: FormOptions<T>) {
 	// https://github.com/hybridly/hybridly/issues/23
 	// TODO: explore unique/automatic key generation
 	const shouldRemember = !!options?.key
@@ -52,9 +61,9 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/** Fields as they were when loaded. */
 	const loaded = safeClone(historyData?.fields ?? options.fields)
 	/** Current fields. */
-	const fields = reactive<T>(safeClone(loaded))
+	const fields = reactive<T>(safeClone(loaded)) as T
 	/** Validation errors for each field. */
-	const errors = ref<Record<keyof T, string>>(historyData?.errors ?? {})
+	const errors = ref<Errors<T>>(historyData?.errors ?? {})
 	/** Whether the form is dirty. */
 	const isDirty = ref(false)
 	/** Whether the submission was recently successful. */
@@ -82,9 +91,9 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Resets the form to its initial values.
 	 */
-	function reset(...keys: (keyof T)[]) {
+	function reset(...keys: P[]) {
 		if (keys.length === 0) {
-			keys = Object.keys(fields)
+			keys = Object.keys(fields) as P[]
 		}
 
 		keys.forEach((key) => {
@@ -95,9 +104,9 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Clear the form fields.
 	 */
-	function clear(...keys: (keyof T)[]) {
+	function clear(...keys: P[]) {
 		if (keys.length === 0) {
-			keys = Object.keys(fields)
+			keys = Object.keys(fields) as P[]
 		}
 
 		keys.forEach((key) => {
@@ -177,9 +186,9 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Clears all errors.
 	 */
-	function clearErrors(...keys: string[]) {
+	function clearErrors(...keys: P[]) {
 		if (keys.length === 0) {
-			keys = Object.keys(fields)
+			keys = Object.keys(fields) as P[]
 		}
 
 		keys.forEach((key) => {
@@ -190,25 +199,25 @@ export function useForm<T extends Fields = Fields>(options: FormOptions<T>) {
 	/**
 	 * Checks if the given keys are dirty in the form.
 	 */
-	function hasDirty(...keys: (keyof T)[]) {
+	function hasDirty(...keys: P[]) {
 		if (keys.length === 0) {
 			return isDirty.value
 		}
 
-		return keys.some((key) => !isEqual(toRaw(fields[key]), toRaw(initial[key])))
+		return keys.some((key) => !isEqual(toRaw(getByPath(fields, key)), toRaw(getByPath(initial, key))))
 	}
 
 	/**
 	 * Clears the given field's error.
 	 */
-	function clearError(key: string) {
+	function clearError(key: P) {
 		unsetPropertyAtPath(errors.value, key)
 	}
 
 	/**
 	 * Sets current errors.
 	 */
-	function setErrors(incoming: Record<string, string>) {
+	function setErrors(incoming: Errors<T>) {
 		clearErrors()
 		Object.entries(incoming).forEach(([path, value]) => {
 			setValueAtPath(errors.value, path, value)

